x-pack/plugins/apm/public/components/shared/DiscoverButtons/DiscoverButton.tsx

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get } from 'lodash';
+import React from 'react';
+import { StringMap } from 'x-pack/plugins/apm/typings/common';
+import {
+  getAPMIndexPattern,
+  ISavedObject
+} from '../../../services/rest/savedObjects';
+import { KibanaLink } from '../../../utils/url';
+
+interface Props {
+  query: StringMap;
+}
+
+interface State {
+  indexPattern?: ISavedObject;
+}
+
+export class DiscoverButton extends React.Component<Props, State> {
+  public state = {};
+  public async componentDidMount() {
+    const indexPattern = await getAPMIndexPattern();
+    this.setState({ indexPattern });
+  }
+
+  public render() {
+    const { query, children, ...rest } = this.props;
+    const id: string = get(this.state, 'indexPattern.id');
+
+    if (!query._a.index) {
+      query._a.index = id;
+    }
+
+    return (
```

Comment:
If we don't have the index pattern ID yet, should we return null? Or somehow disable the link? It's an edge case I suppose but right now if you click the button before the `getAPMIndexPattern` has resolved, it'll still have the no specified index pattern bug…

Reply:
Good point. I don't think we should return `null` because that will make the ui "glitch/jump" as soon as the button is loaded. Adding disabled might be a good idea though.

Reply:
Okay, ended up not disabling the button. The reason is, that there are (at least) two ways to get the APM index pattern:
1) Use the Getting Started import button (recommended)
2) Create it manually by going to Index Pattern management

If the user chooses option 2 we will not be able to find that index pattern (we have no idea what it's called or what the pattern is. Could be `apm-7*-2018*` or something. That would still work in Discover, so we shouldn't disable the button even though we cannot find the index pattern.

Reply:
... index patterns considered harmful ...

lol ok sounds good, I'm on board with not disabling and I don't think it's worth the effort to put the button in a loading state while the request is in progress